### PR TITLE
fix: create schemas at topic creation

### LINKED ISSUE #4219: Create Avro sink schema at time of stream/table creation
**Is your feature request related to a problem? Please describe.**

Currently, when a user creates an Avro sink stream/table, KSQL does not register the Avro schema for the new topic to Schema Registry until the first record is processed. (This is because KSQL relies on the `KafkaAvroSerializer` to automatically register the schema to Schema Registry, and the serializer is only invoked when a record is processed.) However, this leads to unintuitive behavior such as in https://github.com/confluentinc/ksql/issues/3883:
- user issues a query that creates a sink stream with Avro format
- user wishes to define a KSQL table on the sink topic (from the stream above), but the request fails due to `Avro schema for message values on topic <topic name> does not exist in the Schema Registry` 
- a record is sent to through the Streams app defined by the query
- user reissues the create table statement, and this time it succeeds

**Describe the solution you'd like**

KSQL should register the Avro schema with Schema Registry at the time the sink stream/table is defined, rather than waiting until the first record is processed.

**Describe alternatives you've considered**

N/A

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by agavra**
fixes #4219 

### Description 

To be honest, I don't _love_ the way this problem is solved in this PR but I can't come up with anything better. Basically, when we create a new topic in the `TopicCreateInjector` we also register the schema with schema registry. This is easy for `C*` statements but is much more difficult for `C*AS` statements as we don't actually know the schema until we build the logical plan. 

It is important, however, to create the schema _before_ we enqueue onto the command topic so that we don't get into the situation where we re-create schemas while replaying the command topic.

### Testing done 
Unit test and local testing (note no data running through):

```
ksql> CREATE STREAM foo (id VARCHAR) WITH (value_format='avro', partitions=1, kafka_topic='foo');

 Message
----------------
 Stream created
----------------
ksql> CREATE STREAM bar WITH (value_format='avro', kafka_topic='foo');

 Message
----------------
 Stream created
----------------
ksql> DESCRIBE bar;

Name                 : BAR
 Field   | Type
-------------------------------------
 ROWTIME | BIGINT           (system)
 ROWKEY  | VARCHAR(STRING)  (system)
 ID      | VARCHAR(STRING)
-------------------------------------
For runtime statistics and query details run: DESCRIBE EXTENDED <Stream,Table>;
```

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/engine/QueryEngine.java [hunk 1]
```diff
@@ -42,7 +42,7 @@
 import org.slf4j.LoggerFactory;
 
 // CHECKSTYLE_RULES.OFF: ClassDataAbstractionCoupling
-class QueryEngine {
+public class QueryEngine {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/engine/QueryEngine.java]: I don't think we should make this public.  It's an internal impl detail of the engine and should IMHO remain that way.

If you just parse, prepare + execute the query against a sandbox of the engine, (as the rest api normally does), then you'll get back a query, and you can get the schema from that query.

Yes, this is executing some bits we don't need. But it maintains better decoupling in the code IMHO.

If we _really_ want to expose logical plans, then we should expose them _through_ the engine.

File: ksql-engine/src/main/java/io/confluent/ksql/topic/TopicCreateInjector.java [hunk 2]
```diff
@@ -53,18 +63,25 @@
 
   private final KafkaTopicClient topicClient;
   private final MetaStore metaStore;
+  private final SchemaRegistryClient srClient;
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/topic/TopicCreateInjector.java]: I'd personally handle this as a separate injector.  Keep this injector focused on topic creation.

I can see that topic creation is _related_ to schema creation - but the injectors are chained anyway.

File: ksql-engine/src/test/java/io/confluent/ksql/topic/TopicCreateInjectorTest.java [hunk 3]
```diff
@@ -425,6 +440,74 @@ public void shouldCreateMissingTopicWithDefaultCleanupPolicyForWindowedTables()
...
+    config = new KsqlConfig(ImmutableMap.of());
+    givenStatement("CREATE STREAM x (f1 VARCHAR) WITH(kafka_topic='expectedName', value_format='AVRO', partitions=1);");
+    when(builder.build()).thenReturn(new TopicProperties("expectedName", 10, (short) 10));
+
+    // When:
+    injector.inject(statement, builder);
+
+    // Then:
+    verifyNoMoreInteractions(schemaRegistryClient);
+  }
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/topic/TopicCreateInjectorTest.java]: I think this is missing some tests around what happens if schema registration fails...

**Comment by agavra** [comment]
@big-andy-coates - thanks for the review! You bring up a good point about the race conditions. Some comments inline (not necessarily in the order you typed):

> If one node starts the transaction and registers the schema, but another node then init's transactions before the first node produces and commits, then (I believe) the first node's transaction is aborted. So we have a schema registered, but not data source. When the data-source creation is retried, sources in the metastore may have changed, meaning the source now has a different, potentially incompatible schema.

Since the likelihood of this happening is very small (multiple users trying to register the same source with different schemas on different nodes), I'm trying to see if it's worth fixing as part of this PR (I don't think it's trivial at all, see below). That basically means we need to trade off between the existing behavior and this bug. IMO, this fix is worth having even with the knowledge that the race condition exists.

> I think the fix here would be to delete the schema from SR if the transaction is aborted.  ... Maybe injectors need an undo method?. 

I don't think this is sufficient. Imagine the scenario where the race condition actually registers a compatible schema. In this case, the second register call would succeed but the transaction fails. Then we delete the valid schema as part of the undo.

Also as an implementation, it's not trivial as it would require to make injectors stateful. 

> Finally, as @rayokota stated, schema creation in the SR is idempotent. So... maybe a better design, (at least until we move to a metastore db), would be to register the schema when executing the C*AS statements after reading them from the command topic? This would ensure we're only ever registering a schema with the correct columns. Or course, we'd need to handle the registration failure scenario there, but we already check evolvability on the rest-server side, so chance of this are slim. If it does fail, I think it would be ok to put the query into a failed state and log an error message. User could then fix the issue, e.g. delete an old / bad schema, and restart to retry.

Addressing this first as this could have otherwise have solved our problem. As mentioned in the description: "It is important to create the schema before we enqueue onto the command topic so that we don't get into the situation where we re-create schemas while replaying the command topic." We'd run into the same issues we used to get with topic creation during replay.

> Switching to a metastore db model would potentially help here, as the single-writer pattern it would utilise would give us a single node responsible for performing such one-of tasks, without any race conditions.

I've been advocating for a single-writer independent of swithcing the to metastore db model 😉 and I think that's the best fix for lots of these types of race conditions. With the work done to route pull queries to other nodes, I imagine this may not even be that difficult.

File: ksql-engine/src/main/java/io/confluent/ksql/engine/QueryEngine.java (hunk 1 above)
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/engine/QueryEngine.java]: I'm happy with your suggestion and will do that.

File: ksql-engine/src/main/java/io/confluent/ksql/topic/TopicCreateInjector.java (hunk 2 above)
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/topic/TopicCreateInjector.java]: At first I had tried doing that and now I can't remember why I didn't end up going that route. I'll go ahead and do that.

**Comment by big-andy-coates** [comment]
> Addressing this first as this could have otherwise have solved our problem. As mentioned in the description: "It is important to create the schema before we enqueue onto the command topic so that we don't get into the situation where we re-create schemas while replaying the command topic." We'd run into the same issues we used to get with topic creation during replay.

What's wrong with publishing the schema to the SR from each node and on each restart?  As long as the logic is correct, i.e. that we only do this for C*AS commands that will be started, then this would work as far as I can tell.  

This would also avoid the race condition.

If there is some issue with this approach, can you explain it?

Sure, there's much more scope for bugs with this approach, (the whole recreation of schemas for terminated queries, etc). However, I'm sure you can avoid those when implementing ;). Plus, when we switch to a single writer pattern we can move away from this approach.

This seems like the _better_ short term approach to me.

**Comment by agavra** [comment]
Opened https://github.com/confluentinc/ksql/issues/4751

**Comment by apurvam** [comment]
Awesome to see this get in. Thanks @agavra for the initiative and @big-andy-coates for the quick review turnaround!

**Pull request merged**